sql/rewards/rewards.go

```diff
@@ -37,17 +34,41 @@ func Revert(db sql.Executor, revertTo types.LayerID) error {
 	return nil
 }
 
-// List rewards from all layers for the coinbase address.
-func List(db sql.Executor, coinbase types.Address) (rst []*types.Reward, err error) {
-	_, err = db.Exec("select layer, total_reward, layer_reward from rewards where coinbase = ?1 order by layer;",
+// ListByCoinbase lists rewards from all layers for the coinbase address.
+func ListByCoinbase(db sql.Executor, coinbase types.Address) (rst []*types.Reward, err error) {
+	_, err = db.Exec("select pubkey, layer, total_reward, layer_reward from rewards where coinbase = ?1 order by layer;",
 		func(stmt *sql.Statement) {
 			stmt.BindBytes(1, coinbase[:])
 		}, func(stmt *sql.Statement) bool {
+			smesherId := types.NodeID{}
+			stmt.ColumnBytes(0, smesherId[:])
 			reward := &types.Reward{
+				SmesherID:   smesherId,
 				Coinbase:    coinbase,
-				Layer:       types.LayerID(uint32(stmt.ColumnInt64(0))),
-				TotalReward: uint64(stmt.ColumnInt64(1)),
-				LayerReward: uint64(stmt.ColumnInt64(2)),
+				Layer:       types.LayerID(uint32(stmt.ColumnInt64(1))),
+				TotalReward: uint64(stmt.ColumnInt64(2)),
+				LayerReward: uint64(stmt.ColumnInt64(3)),
+			}
+			rst = append(rst, reward)
+			return true
+		})
+	return
+}
+
+// ListBySmesherId lists rewards from all layers for the smesher ID.
+func ListBySmesherId(db sql.Executor, smesherID types.NodeID) (rst []*types.Reward, err error) {
+	_, err = db.Exec("select coinbase, layer, total_reward, layer_reward from rewards where pubkey = ?1 order by layer;",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, smesherID[:])
+		}, func(stmt *sql.Statement) bool {
+			coinbase := types.Address{}
+			stmt.ColumnBytes(0, coinbase[:])
+			reward := &types.Reward{
+				SmesherID:   smesherID,
+				Coinbase:    coinbase,
+				Layer:       types.LayerID(uint32(stmt.ColumnInt64(1))),
+				TotalReward: uint64(stmt.ColumnInt64(2)),
+				LayerReward: uint64(stmt.ColumnInt64(3)),
```

Comment:
this two functions can be written without copying repetition

Reply:
you mean using generics? it's going to be messy. or maybe I don't understand.

Reply:
I think I see what you mean, let me try

Reply:
this is the best I can do, still not sure it's better than what was there before, have a look